### PR TITLE
[WIP] pkg/restmapper: use exponential backoff with DynamicRESTMapper calls

### PR DESCRIPTION
**Pull request by estroz**
**Description of the change:** add exponential backoff to `DynamicRESTMapper.reloadOnErr()` calls.

**Motivation for the change:** see [this discussion](https://github.com/operator-framework/operator-sdk/pull/1329#discussion_r275966907) from #1329.

### CODE REVIEW

File: pkg/restmapper/dynamicrestmapper.go [hunk 1]
```diff
@@ -54,14 +88,19 @@ func (drm *DynamicRESTMapper) reload() error {
...
+// reloadOnError uses an exponential backoff mechanism to rate limit reloads.
 func (drm *DynamicRESTMapper) reloadOnError(err error) bool {
 	if _, matches := err.(*meta.NoKindMatchError); !matches {
+		drm.backoff.reset()
 		return false
 	}
 	err = drm.reload()
 	if err != nil {
+		// TODO(estroz): HandleError uses a rudimentary backoff by default.
+		// Should we remove it completely or substitute the default backoff
```
Comment (theishshah) [inline pkg/restmapper/dynamicrestmapper.go]: I think it makes sense to have the simple back off as a fall-back before failing completely

File: pkg/restmapper/dynamicrestmapper.go [hunk 2]
```diff
@@ -15,30 +15,64 @@
...
 	"k8s.io/apimachinery/pkg/runtime/schema"
 	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/client-go/discovery"
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/restmapper"
 )
 
+// TODO(estroz): do we want to return a wait.ErrWaitTimeout if backoff duration
+// reaches a maximum?
```
Comment (AlexNPavel) [inline pkg/restmapper/dynamicrestmapper.go]: I would say yes. Backoff duration hitting its maximum is, in the end, a wait timeout.

**Comment by estroz** [comment]
@joelanford that would be better UX, especially if this is being used in a reconcile loop. We could use the [token bucket](https://godoc.org/golang.org/x/time/rate#Limiter) approach: if the limiter `Allow`s an event to happen, reload; if not, return an `ErrRateLimited`. We can expose a method on `DynamicRESTMapper.Delay()` to get the time to wait if `ErrRateLimited` is returned, or expose a field on `ErrRateLimited` with that value.

**Comment by joelanford** [comment]
@estroz I was thinking the same think about the reconcile loop issue.

@DirectXMan12 This is definitely related to the conversation going in https://github.com/kubernetes-sigs/controller-runtime/issues/321. Any thoughts on the rate-limiting approach?

**Comment by estroz** [comment]
Instead of making two PR's with these changes (one for the SDK, one to upstream `DynamicRESTMapper`), one has been made to controller-runtime (https://github.com/kubernetes-sigs/controller-runtime/pull/554) and a new PR will remove `pkg/restmapper` from the SDK and make necessary dependency changes once the former is merged and a release made.

**Pull request closed without merging**